### PR TITLE
Nullreferenceexception in new azure ad service principal#9478

### PR DESCRIPTION
**Pull request by VeryEarly**
## Checklist

- [ ] I have read the [_Submitting Changes_](../CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../CONTRIBUTING.md)
- [ ] The title of the PR is clear and informative
- [ ] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [ ] The PR does not introduce [breaking changes](../documentation/breaking-changes/breaking-changes-definition.md)
- [ ] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
    - [ ] the markdown help files have been regenerated using the commands listed [here](../documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

File: src/Resources/Resources/ActiveDirectory/Cmdlets/NewAzureADServicePrincipalCommand.cs [hunk 1]
```diff
@@ -383,7 +391,12 @@ private void CreateSimpleServicePrincipal()
 
         private bool SkipRoleAssignment()
         {
-            return this.IsParameterBound(c => c.SkipAssignment) || (!this.IsParameterBound(c => c.Role) && !this.IsParameterBound(c => c.Scope));
+            return this.IsParameterBound(c => c.SkipAssignment) || (!this.IsParameterBound(c => c.Role) && !this.IsParameterBound(c => c.Scope) && !HasSubscription());
+        }
+
+        private bool HasSubscription()
+        {
+            return DefaultContext.Subscription != null && DefaultContext.Subscription.Id != null ;
```
Comment (markcowl) [inline src/Resources/Resources/ActiveDirectory/Cmdlets/NewAzureADServicePrincipalCommand.cs]:
```suggestion
            return DefaultContext.Subscription?.Id != null ;
```

**Comment by VeryEarly** [comment]
> One minor suggestion. It would be nice to replumb the logic of this so that we didn't have to check for roe assignment skipping more than once, but changing minimum code mitigates risk.
> 
> What are your thoughts about testing this? Seems like we should add a test to prevent us from regressing in this scenario. We can add automation after the release if you file an issue and place it in the next milestone.

I reduce one of the usages of "SkipRoleAssignment", there are still two left and seems not likely to remove either one of them.

I created test case for it but test scenario required a connection to a service principal without subscription which I cannot find how to set up in record mode.

Do you have any idea about these?

**Comment by VeryEarly** [comment]
> > One minor suggestion. It would be nice to replumb the logic of this so that we didn't have to check for roe assignment skipping more than once, but changing minimum code mitigates risk.
> > What are your thoughts about testing this? Seems like we should add a test to prevent us from regressing in this scenario. We can add automation after the release if you file an issue and place it in the next milestone.
> 
> I reduce one of the usages of "SkipRoleAssignment", there are still two left and seems not likely to remove either one of them.
> 
> I created test case for it but test scenario required a connection to a service principal without subscription which I cannot find how to set up in record mode.
> 
> Do you have any idea about these?

git issue, close this pull request, will try fix them and open another one

**Pull request closed without merging**